### PR TITLE
AssertionError when checking dtype on multiple output model

### PR DESCRIPTION
**Pull request by joomyjoo**
Fixed using listify()[0] on self.pred before calling .dtype. Made changes to the ipynb notebooks for testeq-ing self.x and self.pred dtypes for multiple input and multiple output models (the changes to the jupyter notebook does not seem necessary if the example used in the ipynb is going to be one with a single input and single output.

### CODE REVIEW

**Comment by joomyjoo** [comment]
Sry for the needlessly long title.
A merge request for fixing [https://github.com/fastai/fastai/issues/3202](url)

**Comment by joomyjoo** [comment]
This is a nbdev-successful pull request for https://github.com/fastai/fastai/pull/3200

**Comment by joomyjoo** [comment]
In the callback/fp16.py, .dtype is called on self.preds
However, when the model has two or more outputs, self.preds is a tuple of tensors not tensor
This results in an assertion error when self.preds.dtype is called.

To reproduce, train a model with two or more outputs with mixed precision.
"""
/usr/local/lib/python3.6/dist-packages/fastai/callback/fp16.py in after_pred(self)
20 def before_batch(self): self.autocast.enter()
21 def after_pred(self):
---> 22 if self.pred.dtype==torch.float16: self.learn.pred = to_float(self.pred)
23 def after_loss(self): self.autocast.exit()
24 def before_backward(self): self.learn.loss_grad = self.scaler.scale(self.loss_grad)

AttributeError: 'tuple' object has no attribute 'dtype'
"""

Expected Behavior:
The dtype check should be done on one of the outputs, regardless the number of outputs of a model.
Checking all outputs for dtype should not be necessary, because outputs should be of the same dtype.

Fix:
Using listify(self.preds)[0].dtype instead of self.preds[0] in callback/fp16.py
should resolve this issue.

File: nbs/18_callback.fp16.ipynb [hunk 1]
```diff
@@ -770,8 +770,8 @@
     "class TestBeforeMixedPrecision(Callback):\n",
     "    order=-55\n",
     "    def before_fit(self): test_eq(first(self.model.parameters()).dtype, torch.float32)\n",
-    "    def before_batch(self): test_eq(self.x.dtype, torch.float32)\n",
-    "    def after_pred(self): test_eq(self.pred.dtype, torch.float16)\n",
+    "    def before_batch(self): test_eq(listify(self.x)[0].dtype, torch.float32)\n",
```
Comment (jph00) [inline nbs/18_callback.fp16.ipynb:773]: Not sure we need to change the test do we?

**Review by jph00 — CHANGES_REQUESTED** [review]
Thanks for the PR. Can you please add a small simple test to the notebook to show the behavior this change is exercising?

File: nbs/18_callback.fp16.ipynb (hunk 1 above)
Comment (joomyjoo) [reply nbs/18_callback.fp16.ipynb:773]: Yup, we do not need to change the tests.

**Comment by joomyjoo** [comment]
A clean edit of the jupyter notebook with a multioutput example is done [here](https://github.com/fastai/fastai/pull/3217).

Will close this pull request

**Pull request closed without merging**